rust/chains/hyperlane-ethereum/src/mailbox.rs

```diff
@@ -334,10 +339,66 @@ where
             mailbox_address: self.address(),
             mailbox_domain: self.domain.id(),
             root: root.into(),
-            index,
+            index
         })
     }
 
+    #[instrument(level = "debug", err, ret, skip(self))]
+    #[allow(clippy::needless_range_loop)]
+    async fn tree(&self, lag: Option<NonZeroU64>) -> ChainResult<IncrementalMerkle> {
+        let lag = if let Some(lag) = lag { lag.get() } else { 0 }.into();
+
+        let block_number = self
+            .provider
+            .get_block_number()
+            .await
+            .map_err(ChainCommunicationError::from_other)?
+            .saturating_sub(lag)
+            .into();
+
+        let expected_root = self
+            .contract
+            .root()
+            .block(block_number)
```

Comment:
Kinda confused by the `.root().block(bn).call()` here. Can you help me understand what is happening? Not sure if this needs docs or not, I am just unfamiliar.